### ISSUE TITLE
Undocumented and non-configurable memoization feature on apollo-datasource-rest

### ISSUE THREAD

**Issue by Jaxolotl** [apollographql/apollo-server#6603]
## Package name and version
`apollo-datasource-rest V3.6.1`

## Expected behavior 
- I'm able to control whether to enable memoization or not
- If I opt for not using memoization then all my GET requests are executed
- I expect to find the behavior documented somewhere (on the readme file and/or on the apollo server "datasources" documentation
- I am able to specifically decide how to cache the requests (e.g headers, redis, etc)

# Actual behavior
- All [GET requests will be memoized](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-datasource-rest/src/RESTDataSource.ts#L269) no matter you want it or not
- I'm unable to opt out from memoization
- No documentation is available, you have to dig into the source code to find it or accidentally find a clue on a [Khalil Stemmler's blog](https://khalilstemmler.com/blogs/graphql/how-apollo-rest-data-source-caches-api-calls/)
- The package is adding a hidden non configurable caching layer bypassing any other standard caching mechanism

## Description of the case
The datasource will keep a memoized version of all the GET requests until a POST request is made using the same URI or you programmatically delete the entry key (`this.memoizedResults.delete(<MY_KEY>)`) or clear all the entries (`this.memoizedResults.clear()`).

 Not being documented and not having and opt-out configuration for the feature will cause a data inconsistency problem whenever you only make get requests to a system you have just read-only access (if another system updates the data you will never get the new data)

## temporary workaround
You can always do the following to workaround the problem on your datasource derived class but is not a solution
```
export abstract class MyExtendedRestDataSource extends RESTDataSource {
  constructor() {
    super();
  }

  override willSendRequest(request: RequestOptions) {
    this.memoizedResults.clear(); // brute-force clearing the entries before every request
  }
}
```

## Steps to repro
The package code is simple enough and self explanatory to be able to repro with a unit test.

## Extra notes
I saw a couple tickets that might be related to this problem but they were described as generic caching issues without providing a detailed explanation of the problem. Hope you find it useful

https://github.com/apollographql/apollo-server/issues/5236
https://github.com/apollographql/apollo-server/issues/4068

**Comment by Jaxolotl** [comment]
Hi @smyrick thank you for the quick reply!
Here my 2 cents:
### for the hook points:
I'm not really sure how opening the cache implementation for extension/overriding would impact on the project. Personally I wouldn't go in that direction and cannot imagine a scenario where caching a non query verbs ( command-like verb like PUT, POST, DELETE) would benefit from caching

### for the config value:
That sounds less intrusive, although I'd suggest something like this:

```
    if (cacheGetRequests !== false) {
        return performRequest();
    }

    if (request.method === 'GET') {
      let promise = this.memoizedResults.get(cacheKey);
      if (promise) return promise;

      promise = performRequest();
      this.memoizedResults.set(cacheKey, promise);
      return promise;
    } else {
      this.memoizedResults.delete(cacheKey);
      return performRequest();
    }
```
That way we do an earlier return with a simpler condition easy to test and avoiding dealing with the if/else statements. If we add it to the `if (request.method === 'GET' && cacheGetRequests !== false) {` then we should deal with the else deleting a non existing key from the cache all the time

## extra thoughts
I think the feature has a very valid purpose and it would be great to have a better control over it. Switching it on and off might not be a good approach, what if we add a time value, e.g. if 2 GET requests with the same signature have less than 500ms difference then return the memoized promise, otherwise make the request and update the memoized entry, that way you can switched on/off AND you can switched on and define a max life span.
Any thoughts?

**Comment by smyrick** [comment]
I like the idea of adding more control to the caching features but I also want to make sure we don't change the existing behavior of the library. Even if we agree it needs to be updated, having the results potential flushed from the cache would be a breaking change in logic (and require a major version) and someone today might be using this as is.

I totally missed this before but there already is an open in the [cache set](https://github.com/apollographql/apollo-server/blob/7933b26f0025590c22ac3d7aa0868dfc576446b8/packages/apollo-server-core/src/utils/UnboundedCache.ts#L21-L30) to pass in a `ttl` option which just compares to `Date.now()`. So if we made that open to pass in we could customize that:

Default
```js
// Class props
this.cacheEnabled = true;
this.cacheTTL = null;

// Logic
if (this.cacheEnabled) {
  let promise = this.memoizedResults.get(cacheKey);
  if (promise) return promise;

  if (request.method === 'GET') {
    promise = performRequest();
    this.memoizedResults.set(cacheKey, promise, { ttl: cacheTTL });
    return promise;
  } else {
    this.memoizedResults.delete(cacheKey);
    return performRequest();
  }
} else {
  return performRequest();
}
```

**Comment by smyrick** [comment]
@Jaxolotl This is now released in apollo-datasource-rest 3.7
